### DISCUSSION TITLE
Matching rules and priorities for request endpoints

### DISCUSSION THREAD

**Discussion by NC-Cj** [fastapi/fastapi#10942] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.put('/devices/{group_id:str}/{device_id:str}')
def update_device(group_id: str,
                  device_id: str):
    return 'update_device function'


@app.post('/devices/models')
def device_add_models(data: str):
    return "device_add_models function"


@app.put('/devices/models/loading')
def device_load_models(data: str):
    return 'device_load_models function'
```

### Description

request `PUT /devices/models/loading`endpoints, u thought you'd get response：`device_load_models function`, but we got it response: `update_device function`

### Operating System

Windows

### FastAPI Version

0.97.0

### Pydantic Version

1.10.12

### Python Version

3.10

### Additional Context

no

**Comment by NC-Cj** [comment]
Many frameworks adhere strictly to the rules of exact matching, which I didn't find in fastapi, and I was disappointed that it was beyond the developer's understanding😫

**Comment by Kludex** [comment, marked as answer]
The order matters. The first endpoint matched by the rule will be called. It's written somewhere in the documentation.

Move the `@app.put('/devices/models/loading')` to the top, and it will work as you expect.

This is not a FastAPI issue problem anyway. It's Starlette who does the routing.
